internals/overlord/certstate/manager.go

```diff
@@ -0,0 +1,429 @@
+	c.x509KeyPairsLock.Lock()
+	defer c.x509KeyPairsLock.Unlock()
+
+	var selected *x509KeyPair
+
+	// Try to find an existing valid X509
+	for i, keypair := range c.x509KeyPairs {
+		if !isCertExpired(keypair.Certificate) {
+			selected = &c.x509KeyPairs[i]
+			break
+		}
+	}
+
+	// Generate a keypair if no one exists.
+	if selected == nil {
+		selected, err := c.newX509ECP256Keypair()
+		if err != nil {
+			return nil, fmt.Errorf("cannot generate X509 keypair: %w", err)
+		}
+		if err := c.persistX509Keypair(selected); err != nil {
+			return nil, fmt.Errorf("cannot persist X509 keypair: %w", err)
+		}
+
+		// Notify subscribers of the new certificate in use
+		c.callChangeListeners(selected.Certificate)
+	}
+
+	pemCert, err := x509CertToPEM(selected.Certificate)
+	if err != nil {
+		return nil, err
+	}
+	pemKey, err := privateKeyToPEM(selected.Private)
+	if err != nil {
+		return nil, err
+	}
+	tlsCert, err := tls.X509KeyPair(pemCert, pemKey)
+	if err != nil {
+		return nil, err
+	}
+	return &tlsCert, err
+}
+
+// CertChangedFunc is the function type used by AddChangeListener.
+type CertChangedFunc func(cert *x509.Certificate)
+
+// AddChangeListener adds f to the list of functions that are called
+// whenever a cert change event took place (for example, when an
+// expired certificate gets replaced).
+func (c *CertManager) AddChangeListener(f CertChangedFunc) {
+	c.changeListeners = append(c.changeListeners, f)
+}
+
+func (c *CertManager) callChangeListeners(cert *x509.Certificate) {
+	if cert == nil {
+		// Avoids if statement on every deferred call to this method (we
+		// shouldn't call listeners when the operation fails).
+		return
+	}
+	for _, f := range c.changeListeners {
+		f(cert)
+	}
+}
+
+// newX509ECP256Keypair generates an elliptic P256 private key and an
+// X509 self-signed public certificate (signed with the private key). The x509
+// certificate includes both the public key and the signature, which allows
+// the certificate signature to be verified with the public key inside the
+// certificate. This allows a client side copy of the certificate to be used
+// as certificate authority for future TLS sessions with the server.
+func (c *CertManager) newX509ECP256Keypair() (*x509KeyPair, error) {
+	// Get the highest available order
+	order := 1
+	certCount := len(c.x509KeyPairs)
+	if certCount > 0 {
+		// Find the next available order
+		order = c.x509KeyPairs[certCount-1].Order + 1
+		if order > 999 {
+			return nil, fmt.Errorf("cannot process order number %v (valid range is 001-999)", order)
+		}
+	}
+
+	// Valid date range.
+	notBefore := time.Now()
+	notAfter := notBefore.Add(365 * 24 * time.Hour)
+	keypair, err := generateX509ECP256Keypair(notBefore, notAfter)
+	if err != nil {
+		return nil, err
+	}
+	keypair.Order = order
+	return keypair, nil
+}
+
+func (c *CertManager) persistX509Keypair(keypair *x509KeyPair) error {
+	return writeX509Keypair(keypair, c.tlsDir)
```

Comment:
Let's inline this function call (drop `writeX509Keypair()` in favor of `persistX509Keypair()`), since the former is only called here and in the test setup (and this test setup call can be replaced by `persistX509Keyapir(*x509KeyPair)`